### PR TITLE
handling http semantic convention breaking changes in v1.23.0

### PR DESCRIPTION
**Pull request by wangzlei**
**Description:** <Describe what has changed.>

`awsxrayexporter` is able to handle both old and new versions of http span attributes, support translate new http attributes to xray segment fields by following http [server span](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-server) and [client span](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-client) spec.

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30935

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>
https://opentelemetry.io/blog/2023/http-conventions-declared-stable/

### CODE REVIEW

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (mxiamxia) [inline exporter/awsxrayexporter/internal/translator/http.go:65]: are you going to update the new convention keys (in urlParts) later in a separate PR?

**Review by thpierce — COMMENTED** [review]
Didn't review tests, will make time to review tmrw.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (bryan-aguilar) [reply exporter/awsxrayexporter/internal/translator/http.go:65]: I don't know if that's necessary, it looks like the construct logic relies on the non-stable keys. The PR continues to use the non stable attributes and keys even if the stable attribute was present in the span. It would probably make sense to switch over to primarily using the stable semantic conventions when when the new `semconv` packages are released.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (bryan-aguilar) [reply exporter/awsxrayexporter/internal/translator/http.go:216]: Why did this logic need to get extracted out of the nested conditional below?

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (mxiamxia) [reply exporter/awsxrayexporter/internal/translator/http.go:65]: make sense!

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:28]: Yes, when Collector has full http attribute keys in semconv v1.23.0, we will replace them by constants in semconv v1.23.0

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:71]: Good question. Because http semantic conventions have been changed many times, v1.20 to v1.23.0 are just the latest changed. The change history is:
`http.host`(deprecated in v1.13.0) ->  `net.host.name`(deprecated) -> `server.address`

The awsxrayexporter translation is stale for a long time, this PR includes more changes than v.1.23.0.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:78]: It is deprecated in semconv v.1.19.0

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:65]: Existing function `constructClientURL` and `constructServerURL` still use old conventions, we will update them to use new conventions when the stable semconv are involved in Collector.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:216]: It is a bug fix. The bug is detected by test case `TestServerSpanWithSchemeNamePortTargetAttributesStable`

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:96]: `net.peer.port` is deprecated in v1.23.0, use `server.address` on client spans and `client.address` on server spans. That means if span has attribute `client.address` we can follow the existing 'net.peer.port' translation logic.

This case no need to cover `server.address` because it is already covered by another case:
```
case conventions.AttributeHTTPHost, AttributeServerAddress:
			urlParts[conventions.AttributeHTTPHost] = value.Str()
			hasHTTP = true
			hasHTTPRequestURLAttributes = true
```

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:76]: Good question. URL query may contain user sensitive information, so xray would mask URL query even there is. By this opportunity we will only show URL path for HTTP url.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:56]: Nice call out! `client.address` contains 2 cases: 1) http; 2) network.peer.address. I will add a logic to distinguish them.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [inline exporter/awsxrayexporter/internal/translator/http.go:112]: I don't really understand this. Can you. explain? I'm trying to follow using the migration guide but maybe there is information elsewhere?

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [inline exporter/awsxrayexporter/internal/translator/http.go:113]: Don't we need to check for `AttributeNetworkPeerAddress`, too?

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [inline exporter/awsxrayexporter/internal/translator/http.go:112]: I would advise elaborating on this comment, it's not abundantly clear why this is true.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [reply exporter/awsxrayexporter/internal/translator/http.go:69]: Please provide evidence of `http.host` -> `net.host.name`. I could only find https://github.com/open-telemetry/opentelemetry-specification/pull/2469, which says `http.host` -> `http.request.header.host`

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [reply exporter/awsxrayexporter/internal/translator/http.go:74]: Not sure I understand your comment. You linked to AWS XRay SDK, are you saying you are doing this to intentionally bring `opentelemetry-collector-contrib` into parity with XRay SDK? If so, should we remove `AttributeHTTPTarget` from this case/attempt to parse out queries?

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [reply exporter/awsxrayexporter/internal/translator/http.go:76]: Deprecated here: https://github.com/open-telemetry/opentelemetry-specification/pull/2469, in favor of `net.host.name`, which is deprecated in favor of `server.address` in current migration guide. Seems accurate.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [reply exporter/awsxrayexporter/internal/translator/http.go:94]: So, I don't think `*.address` is relevant, you probably meant `*.port`. In any case, [here's where `net.peer.port` was deprecated](https://github.com/open-telemetry/opentelemetry-specification/pull/3402/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR63-R65), this is telling us that `net.peer.port` was deprecated in favor of two attributes `server.port` and `client.port`:

> BREAKING: rename `net.peer.name` to `server.address` on client side and to `client.address` on server side, `net.peer.port` to `server.port` on client side and to `client.port` on server side, `net.host.name` and `net.host.port` to `server.name` and `server.port` (since `net.host.*` attributes only applied to server instrumentation)

So in a few cases, we are splitting one attribute into two. What this means is if we were once looking for `net.peer.port`, we now need to look for both `server.port` and `client.port`, unless we have some way of distinguishing if we are in the server or client perspective.

Now, you do have a case for `server.port`, but that doesn't do the same thing:

```
case conventions.AttributeNetHostPort, AttributeServerPort:
  urlParts[conventions.AttributeNetHostPort] = value.Str()
  hasHTTP = true
  if len(urlParts[conventions.AttributeNetHostPort]) == 0 {
    urlParts[conventions.AttributeNetHostPort] = strconv.FormatInt(value.Int(), 10)
  }
```

vs

```
case conventions.AttributeNetPeerPort, AttributeClientPort:
  urlParts[conventions.AttributeNetPeerPort] = value.Str()
  if len(urlParts[conventions.AttributeNetPeerPort]) == 0 {
    urlParts[key] = strconv.FormatInt(value.Int(), 10)
  }
```

What this seems to suggest is that if we see either AttributeServerPort or AttributeClientPort, we need to run the AttributeNetPeerPort logic, and if we see AttributeServerPort, we need to also run the AttributeNetHostPort logic.

Similar logic follows for `net.peer.name`

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (thpierce) [reply exporter/awsxrayexporter/internal/translator/http.go:220]: I would have greatly preferred if you split this out into a separate PR, since it's not related to the overall changes. Can you explain what the bug is and why the fix is correct? Please also update the overview to explain before/after of this bugfix.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:113]: No need. Because there is
```
if !ok && info.Reuqest.ClientIP != nil
```

Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:113]: No need, because both `conventions.AttributeNetPeerIP` and  `AttributeNetworkPeerAddress` are stored in`urlParts[conventions.AttributeNetPeerIP]`, which is the pattern we used for compatibility.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:112]: Please help provide a better wording? `client.address` is either from "network peer address" or from "X-Forwarded-For". If it is from "network peer address", the span has both `client.address` and "network peer address". If there is no "network peer address", the `client.address` must be from "X-Forwarded-For".

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:74]: Correct, the further fix is to `Sanitize` the `query` part from `conventions.AttributeHTTPTarget`, we can address it in next PR.

Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go:74]: Correct. awxrayexporter is parity to xray sdks. We need to sanitize the Query from Target.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go]: Fix this issue by adding `AttributeServerAddress` to urlParts, for building http server span URL later.
```
case AttributeServerAddress:
			urlParts[key] = value.Str()
			hasHTTPRequestURLAttributes = true
```

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go]: We are lucky that `url.full` is required attribute in Http client span, so we don't have to build http url for http client span by `client.address` and `client.port`.
Only `server.address` and `server.port` are needed for building http url for http server span.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go]: We are lucky that url.full is required attribute in Http client span, so we don't have to build http url for http client span by client.address and client.port.
Only server.address and server.port are needed for building http url for http server span.

File: exporter/awsxrayexporter/internal/translator/http.go
Comment (wangzlei) [reply exporter/awsxrayexporter/internal/translator/http.go]: Adding a boolean flag `hasNetPeerAddr`, so, we can confirm the client IP( only if IP exists) is from X-Forwarded-For or not.
```
if !hasNetPeerAddr && info.Request.ClientIP != nil {
		info.Request.XForwardedFor = aws.Bool(true)
	}
```

**Review by mxiamxia — APPROVED** [review]

**Comment by wangzlei** [comment]
@bogdandrutu @atoulme  Would you help us review and merge this PR? 
[AWS CloudWatch](https://github.com/amazon-contributing/opentelemetry-collector-contrib) schedules a release next week, right now we are blocked by this PR.

**Comment by TylerHelmuth** [comment]
Please address lint issues

**Comment by wangzlei** [comment]
> Please address lint issues

@TylerHelmuth addressed! Please take a look.

**Comment by wangzlei** [comment]
@open-telemetry/collector-contrib-approvers Would you help us review and merge this PR? 
[AWS CloudWatch](https://github.com/amazon-contributing/opentelemetry-collector-contrib) schedules a release this week, right now we are blocked by this PR. Appreciate!

**Review by bryan-aguilar — APPROVED** [review]

**Pull request merged**
